gh-126554: ctypes: Correctly handle NULL dlsym values

grgalex opened the pull request:
For dlsym(), a return value of NULL does not necessarily indicate an error [1].

Therefore, to avoid using stale (or NULL) dlerror() values, we must:

 1. clear the previous error state by calling dlerror()
 2. call dlsym()
 3. call dlerror()

If the return value of dlerror() is not NULL, an error occured.

In our case of ctypes, I understand that we subjectively treat a NULL return value from dlsym() as an error, so we must format a special string for that.

[1] https://man7.org/linux/man-pages/man3/dlsym.3.html

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-126554 -->
* Issue: gh-126554
<!-- /gh-issue-number -->

grgalex committed:
ctypes: Correctly handle NULL dlsym values

For dlsym(), a return value of NULL does not necessarily indicate
an error [1].

Therefore, to avoid using stale (or NULL) dlerror() values, we must:

 1. clear the previous error state by calling dlerror()
 2. call dlsym()
 3. call dlerror()

If the return value of dlerror() is not NULL, an error occured.

In our case, we also subjectively treat a NULL return value
from dlsym() as an error, so we format a special string for that.

[1]: https://man7.org/linux/man-pages/man3/dlsym.3.html

Signed-off-by: Georgios Alexopoulos <grgalex42@gmail.com>

ZeroIntensity commented on Misc/NEWS.d/next/C_API/2024-11-07-20-24-58.gh-issue-126554.ri12eb.rst:
```diff
@@ -0,0 +1 @@
+Correctly handle NULL dlsym() values in ctypes.
```
I prefer to not expose technical details in the changelog, it's not particularly useful to users:
```suggestion
Fix rare crash upon system failure in :class:`ctypes.CDLL`.
```

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -967,18 +967,30 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
-        PyErr_SetString(PyExc_ValueError, dlerror());
-#endif
+    char *dlerr;
+    dlerr = dlerror();
```
Formatting nitpick:
```suggestion
    char *dlerr = dlerror();
```

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -967,18 +967,30 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
         return NULL;
     }
 #else
+    dlerror();
```
I'm not a huge fan of this workaround, but I guess it's simple enough. It would be nice to emit a warning (or possibly raise an error early) if this fails. Also, could you add a comment with the issue number explaining why this is needed, so we don't confuse the future readers?

grgalex replied:
ACK, I'll add a comment explaining the rationale.

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -967,18 +967,30 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
+    char *dlerr;
+    dlerr = dlerror();
+    if (dlerr) {
+        PyErr_SetString(PyExc_ValueError, dlerr);
```
I'm worried that there was a second bug here: this assumes that UTF-8 is the default locale. I'm pretty sure that can cause problems, so I *think* we're going to want to use `PyUnicode_DecodeLocale` here, and then pass the result to `PyErr_SetObject`.

grgalex replied:
Should this be a separate issue/investigation, followed by another PR?

In the sense that if we make a decision on this, we'll have to apply throughout.

ZeroIntensity replied:
I'll investigate this a little further. It's probably better to deal with it in another PR. For now, could you add an `// XXX Locale problems?` or something similar?

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3774,19 +3786,33 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
-        PyErr_SetString(PyExc_AttributeError, dlerror());
-#endif
+    char *dlerr;
+    dlerr = dlerror();
```
```suggestion
    char *dlerr = dlerror();
```

ZeroIntensity reviewed (commented):
It's good to see new contributors :)

Unfortunately, this change does need a test. If it's really, *really* difficult to add one, then I guess we'll figure things out, but in general, it's difficult for us to accept even *tiny* bugfixes without proper tests.

For something like this, as long as the test works, it's probably OK. You could do something as hacky as hard-coding a byte string into a test file and unpacking that into a `.so` at runtime, and just disable all the systems that don't work. But it would be preferable if you could find a library that actually triggers this problem in the wild.

grgalex committed:
Update Misc/NEWS.d/next/C_API/2024-11-07-20-24-58.gh-issue-126554.ri12eb.rst

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

grgalex committed:
Update Modules/_ctypes/_ctypes.c

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

grgalex committed:
Update Modules/_ctypes/_ctypes.c

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

grgalex committed:
Add dlerror() comment, handle callproc.c

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
test: Add ctypes null dlsym case

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

ZeroIntensity commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,69 @@
+@unittest.skipUnless(sys.platform.startswith('linux'),
+                     'Test only valid for Linux')
+class TestNullDlsym(unittest.TestCase):
+    def test_null_dlsym(self):
```
Please add a comment with something like `# GH-126554: ...` explaining the issue in brief.

grgalex replied:
Done.

grgalex committed:
Add comment to test, minor wip-debugging prints

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
Remove wip/debug prints

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

ZeroIntensity commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,90 @@
+                # and since we subjectively treat NULL
+                # addresses as errors, we should get
+                # an error.
+                L.foo
```
Can we assert that this is equal to something?

grgalex replied:
The precise thing to assert here is that an exception should be raised immediately after executing this command.

Therefore, I suggest adding a `self.fail()` with a comment, similar to what we do in https://github.com/python/cpython/blob/54c63a32d06cb5f07a66245c375eac7d7efb964a/Lib/test/test_ctypes/test_structunion.py#L347

WDYT?

grgalex replied:
I'm not sure how the `with self.assertRaises(AttributeError) as cm:` block works.

Will it break from this block if an AttributeError is raised? Is this a wrapper around `try/except`?

[edit: Yes, it will break out of the block]

ZeroIntensity replied:
Hmm, actually, leave this be. I'm just not used to seeing attributes alone 😅

efimov-mikhail replied:
This block checks that there was an exception. It will generate new exception in another case.

grgalex replied:
I tried this and verified it works. Pushed another commit with an explicit failure if no attribute error was raised.

grgalex replied:
I hadn't seen your comment, you can see the commit and judge if you want to keep it or not :)

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -967,18 +967,36 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
+    char *dlerr = dlerror();
+    if (dlerr) {
+		// XXX: This assumes that UTF-8 is the default locale.
+		//      Investigate if this can cause problems.
```
Align indentation:
```suggestion
		  // XXX: This assumes that UTF-8 is the default locale.
		  //      Investigate if this can cause problems.
```

grgalex replied:
Right! Forgot to (re)set expandtab on vim!

Did you catch this with some automated tool?

ZeroIntensity commented on Modules/_ctypes/callproc.c:
```diff
@@ -1623,10 +1623,25 @@ static PyObject *py_dl_sym(PyObject *self, PyObject *args)
+	char *dlerr = dlerror();
+    if (dlerr) {
+		// XXX: This assumes that UTF-8 is the default locale.
+		//      Investigate if this can cause problems.
```
And again
```suggestion
		  // XXX: This assumes that UTF-8 is the default locale.
		  //      Investigate if this can cause problems.
```

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3774,19 +3792,39 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
+    char *dlerr = dlerror();
+    if (dlerr) {
+		// XXX: This assumes that UTF-8 is the default locale.
+		//      Investigate if this can cause problems.
```
Same here:
```suggestion
		  // XXX: This assumes that UTF-8 is the default locale.
		  //      Investigate if this can cause problems.
```

ZeroIntensity reviewed (commented):
Getting very close to being ready :)

grgalex committed:
Update Modules/_ctypes/_ctypes.c

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

grgalex committed:
Update Modules/_ctypes/callproc.c

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

grgalex committed:
Update Modules/_ctypes/_ctypes.c

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

grgalex committed:
Explicitly fail if not AttributeError raised

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

efimov-mikhail commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,91 @@
+                # addresses as errors, we should get
+                # an error.
+                L.foo
+                self.fail("AttributeError should have been raised!")
```
This line is not necessary, but I'm okay with keeping it.

efimov-mikhail commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,91 @@
+            # Load the shared library
+            L = CDLL(dstname)
+
+            with self.assertRaises(AttributeError) as cm:
```
We could ommit "as cm" part.

grgalex replied:
How does the
```
            self.assertEqual(str(cm.exception),                                 
                             "function 'foo' not found")
```

part even work, since it's out of the `with ... as cm` block?

I might have misunderstood something here.

ZeroIntensity replied:
You probably want `assertRaisesRegex`

grgalex replied:
No, I mean, how can we access the `cm` object, since we're out of the "with ... as cm" block?

```
            with self.assertRaises(AttributeError) as cm:                       
                L.foo                                                           
                self.fail("AttributeError should have been raised!")            
                                                                                
            self.assertEqual(str(cm.exception),                                 
                             "function 'foo' not found") 
```

ZeroIntensity replied:
Yeah, don't use `assertRaises` for checking an error message, use `assertRaisesRegex`.

grgalex replied:
What I mean is that the code works as is right now. The test passes.

What I'm wondering is *why does it even work*? How come can we access `cm.exception` outside of the `with ... as cm` block?

ZeroIntensity replied:
Because context variables aren't destroyed after the `with` block.

grgalex replied:
Ok, that was a bit embarassing :)

Thanks for the clarification.

So, should we keep it as is and remove the redundant `self.fail("AttributeError should have been raised!")` part, since we already assert that an AttributeError will be raised?

ZeroIntensity replied:
I'd recommend removing the redundant `self.fail`, but I would like it if it used `assertRaisesRegex` instead of manually checking the `str()`.

grgalex replied:
Done and pushed! Looks way cleaner.

efimov-mikhail reviewed (commented):
I can see two additional calls of `dlsym` is code base: from Python/dynload_shlib.c and from `load_dyld_shared_cache_contains_path` in Modules/_ctypes/callproc.c.

IMO, it's better to provide one function for dlsym, since we have 5 different calls.

grgalex committed:
Tidy up Assertion in test

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -967,18 +967,36 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
+    char *dlerr = dlerror();
+    if (dlerr) {
+		  // XXX: This assumes that UTF-8 is the default locale.
+		  //      Investigate if this can cause problems.
```
Sorry, I might have screwed things up earlier with the indentation. (I made the review from my phone.)
There's some mixing of tabs and spaces here, so this should hopefully be the correct suggestion. My bad!
```suggestion
        // XXX: This assumes that UTF-8 is the default locale.
        // Investigate if this can cause problems.
```

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3774,19 +3792,39 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
+    char *dlerr = dlerror();
+    if (dlerr) {
+		  // XXX: This assumes that UTF-8 is the default locale.
+		  //      Investigate if this can cause problems.
```
On second thought, let's not repeat this comment three times though.
```suggestion
```

ZeroIntensity commented on Modules/_ctypes/callproc.c:
```diff
@@ -1623,10 +1623,25 @@ static PyObject *py_dl_sym(PyObject *self, PyObject *args)
+	char *dlerr = dlerror();
+    if (dlerr) {
+		  // XXX: This assumes that UTF-8 is the default locale.
+		  //      Investigate if this can cause problems.
```
```suggestion
```

ZeroIntensity reviewed (commented):
Just one more fix with the comments and then this looks good.

grgalex committed:
Update Modules/_ctypes/_ctypes.c

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

grgalex committed:
Update Modules/_ctypes/_ctypes.c

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

grgalex committed:
Update Modules/_ctypes/callproc.c

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

ZeroIntensity reviewed (dismissed):
Thanks for all your hard work on this, LGTM!

picnixz commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,88 @@
+   https://willnewton.name/uncategorized/using-gnu-indirect-functions/
+
+   Adapted from Michael Kerrisk's answer: https://stackoverflow.com/a/53590014
+*/
```
```suggestion
/* This is a 'GNU indirect function' (IFUNC) that will be called by
   dlsym() to resolve the symbol "foo" to an address. Typically, such
   a function would return the address of an actual function, but it
   can also just return NULL.  For some background on IFUNCs, see
   https://willnewton.name/uncategorized/using-gnu-indirect-functions.

   Adapted from Michael Kerrisk's answer: https://stackoverflow.com/a/53590014.
*/
```

picnixz commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,88 @@
+@unittest.skipUnless(sys.platform.startswith('linux'),
+                     'Test only valid for Linux')
+class TestNullDlsym(unittest.TestCase):
+    """GH-126554: Ensure that we catch NULL dlsym return values
```
In general, we prefer not to have docstrings for tests because when executed in verbose mode, they scramble the output (though I'm not sure if it's the case for *class* docstrings).

picnixz replied:
Update: class docstrings don't mangle the output in verbose mode so it's fine (see https://github.com/python/cpython/pull/126555#discussion_r1835716669).

picnixz commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,88 @@
+class TestNullDlsym(unittest.TestCase):
+    """GH-126554: Ensure that we catch NULL dlsym return values
+
+    In rare cases, such as when using GNU IFUNCs, dlsym(),
+    the C function that ctypes' CDLL uses to get the address
+    of symbols, can return NULL.
+
+    The objective way of telling if an error during symbol
+    lookup happened is to call glibc's dlerror() and check
+    for a non-NULL return value.
+
+    However, there can be cases where dlsym() returns NULL
+    and dlerror() is also NULL, meaning that glibc did not
+    encounter any error.
+
+    In the case of ctypes, we subjectively treat that as
+    an error, and throw a relevant exception.
+
+    This test case ensures that we correctly enforce
+    this 'dlsym returned NULL -> throw Error' rule.
```
Is this summarized on the issue? if so, we don't need to be extra verbose here. Tests are usually not that descriptive in what they do, except if there are some real needs. (For instance, I would only keep the first sentence and drop the rest if this is already known on the issue / PR).

ZeroIntensity replied:
Eh, I'm personally OK with it here. This is a complicated issue and equally complicated test case, I'd rather more info on it than not enough.

picnixz replied:
Ok, I can buy that argument. Though I'd like to know whether this could garble the output in verbose mode (this is something we don't want) in case of a failure/success. If this isn't the case, then I'm also ok (but the description is alsmot longer than the test I think?)

ZeroIntensity replied:
I'll double-check. I do remember seeing docstrings in some of the C API tests, I think.

ZeroIntensity replied:
I think it's ok for class docstrings, it looks fine locally.

picnixz commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,88 @@
+                                 stderr=subprocess.DEVNULL)
+            out, _ = p.communicate()
+        except OSError:
+            raise unittest.SkipTest('gcc, needed for test, not available')
```
```suggestion
            self.skipTest('gcc is missing')
```

picnixz commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,88 @@
+        try:
+            p = subprocess.Popen(['gcc', '--version'], stdout=subprocess.PIPE,
+                                 stderr=subprocess.DEVNULL)
+            out, _ = p.communicate()
```
AFACT, this variable is not used. Can we use someting else to check for the existence of `gcc`, e.g., `subprocess.run` + catch exception?

ZeroIntensity replied:
Would `subprocess.check_call` work?

picnixz replied:
I always forget which one is the easiest to use but I think we can use this (no need for catching exceptions):

```py
retcode = subprocess.call(["gcc", "--version"], 
                          stdout=subprocess.DEVNULL,
                          stderr=subprocess.DEVNULL,
                          shell=True)
if retcode != 0:
    self.skipTest("gcc is missing")
```

grgalex replied:
Done and pushed.

grgalex replied:
Actually, something might be off.

I get a `retcode` of 1, even though `gcc` is present, in the sense that if I remove the check, the lines below compile `libfoo.so` normally, and the test completes.

picnixz replied:
You're right, here's what I got:

```pycon
>>> subprocess.call(["gcc", "--version"])
gcc (SUSE Linux) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

0
>>> subprocess.call(["gcc", "--version"], shell=True)
gcc: fatal error: no input files
compilation terminated.
1
```

picnixz replied:
Ok, so the `shell` parameter may not be necessary. But you need to catch the OSError still...

grgalex replied:
It was the `shell=True` part. I logged the stderr, and I get `gcc: No input specified`.

This means that the `--version` argument was not passed correctly.

I'll investigate. We want to keep the `shell=True` part, but have it work as expected.

picnixz replied:
Correct code:

```py
try:
    retcode = subprocess.call(["gcc", "--version"],
                              stdout=subprocess.DEVNULL,
                              stderr=subprocess.DEVNULL)
except OSError:
    self.skipTest("gcc is missing")
else:
    if retcode != 0:
        self.skipTest("gcc --version failed")
```

picnixz replied:
I don't necessarily think we need the `shell` part since we're anyway on Linux only. Btw, the reason why `--version` failed is because it expected `gcc --version` as a single string and not a list of arguments (though I don't why it seems to incorrectly dispatch the command). So you can also use:

```py
retcode = subprocess.call("gcc --version",
                          stdout=subprocess.DEVNULL,
                          stderr=subprocess.DEVNULL,
                          shell=True)
if retcode != 0:
    self.skipTest("gcc is missing")
```

grgalex replied:
Figured it out.

When using `shell=True`, the command to run must be a single string, passed as the first argument.

Any additional arguments are passed to the shell itself.

So, in this case, we were running the equivalent of:

```
/bin/sh -c 'gcc' --version
```

Quoting the docs [1]:

> On POSIX with shell=True, the shell defaults to /bin/sh. If args is a string, the string specifies the command to execute through the shell. This means that the string must be formatted exactly as it would be when typed at the shell prompt. This includes, for example, quoting or backslash escaping filenames with spaces in them. If args is a sequence, the first item specifies the command string, and any additional items will be treated as additional arguments to the shell itself. That is to say, [Popen](https://docs.python.org/3/library/subprocess.html#subprocess.Popen) does the equivalent of: `Popen(['/bin/sh', '-c', args[0], args[1], ...])`

[1] https://docs.python.org/3/library/subprocess.html#popen-constructor

grgalex replied:
I'll merge into a single string, "gcc --version" and keep `shell=True`.

grgalex replied:
@picnixz I'm on Firefox, and I wasn't seeing your comments in real time for some reason.

So the above, which kinda looks like a conversation, was a monologue on my side...

If we keep the `shell=True`, no need to catch the OSError.

If we remove the `shell=True`, we'll have to catch it.

grgalex replied:
Let's keep `shell=True`, which is in line with the commit I pushed.

picnixz replied:
>  I'm on Firefox, and I wasn't seeing your comments in real time for some reason.

We're both on FF then and I also thought it was a monologue.

grgalex replied:
We reached the same conclusion any way :P

picnixz commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,88 @@
+            out = subprocess.check_output(cmd)
+            self.assertTrue(os.path.exists(dstname))
+            # Load the shared library
+            L = CDLL(dstname)
```
```suggestion
        with tempfile.TemporaryDirectory() as d:
            # Create a C file with a GNU Indirect Function (FOO_C)
            # and compile it into a shared library.
            srcname = os.path.join(d, 'foo.c')
            dstname = os.path.join(d, 'libfoo.so')
            with open(srcname, 'w') as f:
                f.write(FOO_C)
            args = ['gcc', '-fPIC', '-shared', '-o', dstname, srcname]
            p = subprocess.run(args, capture_output=True)
            self.assertEqual(p.returncode, 0, p)
            # Load the shared library
            L = CDLL(dstname)
```

You can simplify the test as follows. No need for more I think.

ZeroIntensity replied:
If we're going to go this route, let's simplify the `'lib%s.so' % 'py_ctypes_test_null_dlsym'` down to just `libpy_ctypes_test_null_dlsym.so` as well.

picnixz replied:
I kept the `%` in order to see the libname itself for debugging but we can also inline the entire filename.

EDIT: Since we are using `foo.c`, I think `libfoo.so` is also fine.

picnixz replied:
I've changed the command to `p = subprocess.run(args, capture_output=True)` so that it captures the output. And if the test fails, then we display `p` directly so that we know what happened (namely if p.returncode != 0).

grgalex replied:
Applied, tested and pushed. Looks very clean.

picnixz commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,88 @@
+import os.path
```
```suggestion
import os
```

picnixz commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -967,18 +967,36 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
 #else
-        PyErr_SetString(PyExc_ValueError, dlerror());
-#endif
+    char *dlerr = dlerror();
```
```suggestion
    const char *dlerr = dlerror();
```

picnixz commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3774,19 +3792,37 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
+	 * Clear the previous value before calling dlsym(),
+	 * to ensure we can tell if our call resulted in an error.
+	 */
+    dlerror();
```
```suggestion
    (void)dlerror();
```

picnixz commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3774,19 +3792,37 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
 #else
-        PyErr_SetString(PyExc_AttributeError, dlerror());
-#endif
+    char *dlerr = dlerror();
```
```suggestion
    const char *dlerr = dlerror();
```

picnixz commented on Modules/_ctypes/callproc.c:
```diff
@@ -1623,10 +1623,23 @@ static PyObject *py_dl_sym(PyObject *self, PyObject *args)
+	 * Clear the previous value before calling dlsym(),
+	 * to ensure we can tell if our call resulted in an error.
+	 */
+	dlerror();
```
```suggestion
	(void)dlerror();
```

picnixz commented on Modules/_ctypes/callproc.c:
```diff
@@ -1623,10 +1623,23 @@ static PyObject *py_dl_sym(PyObject *self, PyObject *args)
-    if (!ptr) {
-        PyErr_SetString(PyExc_OSError,
-                               dlerror());
+	char *dlerr = dlerror();
```
```suggestion
	const char *dlerr = dlerror();
```

grgalex committed:
Update Lib/test/test_ctypes/test_dlerror.py

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

grgalex committed:
Update Lib/test/test_ctypes/test_dlerror.py

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

picnixz commented on Misc/NEWS.d/next/C_API/2024-11-07-20-24-58.gh-issue-126554.ri12eb.rst:
```diff
@@ -0,0 +1 @@
+Fix rare crash upon system failure in :class:`ctypes.CDLL`.
```
```suggestion
Fix error handling in :class:`ctypes.CDLL` objects
which could result in a crash in rare situations.
```

grgalex replied:
Way smoother!

picnixz commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -967,18 +967,36 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
+	 * Clear the previous value before calling dlsym(),
+	 * to ensure we can tell if our call resulted in an error.
+	 */
+    dlerror();
```
```suggestion
    (void)dlerror();
```

grgalex committed:
test: Smoothly check for existence of GCC

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

picnixz commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,89 @@
+        retcode = subprocess.call(["gcc --version"],
+                                  stdout=subprocess.DEVNULL,
+                                  stderr=subprocess.DEVNULL,
+                                  shell=True)
```
```suggestion
        retcode = subprocess.call("gcc --version",
                                  stdout=subprocess.DEVNULL,
                                  stderr=subprocess.DEVNULL,
                                  shell=True)
```

grgalex replied:
Should I also add a short comment that explains why we opt to use a string vs a single-element list?

(We do it to emphasize that only the first elem of the list would get passed as one would expect, the rest of the args are passed to the shell)

picnixz replied:
I don't think it's needed; it's just that we didn't use the API correctly. But if you think it's needed for *future* readers (always keep them in mind), then go for it.

grgalex replied:
Pushed two commits (forgot to add something in the first) that explain this.

If it's too verbose you can chop it.

grgalex committed:
test: Clean up libfoo.so compilation command

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
Update Lib/test/test_ctypes/test_dlerror.py

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

grgalex committed:
Update Modules/_ctypes/_ctypes.c

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

grgalex committed:
Update Modules/_ctypes/_ctypes.c

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

grgalex committed:
Update Modules/_ctypes/_ctypes.c

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

grgalex committed:
Update Modules/_ctypes/callproc.c

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

grgalex committed:
Update Modules/_ctypes/callproc.c

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

grgalex committed:
Update Modules/_ctypes/_ctypes.c

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

grgalex committed:
Update Misc/NEWS.d/next/C_API/2024-11-07-20-24-58.gh-issue-126554.ri12eb.rst

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

grgalex committed:
test: Clarify string vs list in subprocess.run

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
test: Use string in gcc --version

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

picnixz commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,97 @@
+        # To avoid mistakes, pass a single string
+        # (equivalent to a single-elem list),
+        # with the shell command to execute.
+        retcode = subprocess.call("gcc --version",
```
```suggestion
        # Recall: using shell=True is equivalent to: /bin/sh -c 'args[0]' args[1] ...,
        # so we need to pass a list with a single item, namely the command to run.
        retcode = subprocess.call(["gcc --version"],
```

Since we are saying "args[0]" and "args[1]", it's better to keep the list-form. Without the comment, the string form could be fine but with the comment, the list form makes more sense (otherwise you need to mention that it's equivalent to a 1-element list).

grgalex replied:
It's more concise this way. Thanks!

grgalex committed:
Update Lib/test/test_ctypes/test_dlerror.py

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

encukou commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3774,19 +3792,37 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
+	 * to ensure we can tell if our call resulted in an error.
+	 */
+    (void)dlerror();
     address = (PPROC)dlsym(handle, name);
```
This isn't thread-safe, but
- I don't see a way to make it thread-safe, since `dlerror()` &c. can be invoked by non-Python code.
- it should be OK for getting the *message* of an error.

As for the locale, I'd assume that `dlerror` returns ASCII. It certainly cannot support *arbitrary* locales (e.g. UTF-16 doesn't work in “zero-terminated `char *`”).

ZeroIntensity replied:
`dlerror()` is thread-safe (as in, it uses a thread local), at least according to the man pages.

encukou replied:
On Linux, yes, but [POSIX says](https://pubs.opengroup.org/onlinepubs/9699919799.2013edition/functions/dlerror.html): 
> It is implementation-defined whether or not the dlerror() function is thread-safe. A thread-safe implementation shall return only errors that occur on the current thread.


I was wrong elsewhere: an UTF-16 locale is nonsense.
AFAICS, in glibc the message is meant to be translated, but might not be: https://repo.or.cz/glibc.git/blob/HEAD:/elf/dl-lookup.c#l795
So, `PyUnicode_DecodeLocale` & `PyErr_SetObject` would make sense.

ZeroIntensity replied:
Hmm, we have uses of `dlerror` elsewhere, so I'd prefer that we deal with locale and thread safety in another PR. Though, are there systems that we even *support* that can have a racy `dlerror()`? I'm not a fan of ruining performance with lock contention.

picnixz replied:
@ZeroIntensity  I can take care of dealing with locale since you're more informed in threads than me :)

ZeroIntensity replied:
Go for it :)

I think the best approach here would be to add a new internal function so we're not repeating ourselves, e.g. `_PyErr_SetDlerror()`.

encukou commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3774,19 +3792,37 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
-        PyErr_SetString(PyExc_AttributeError, dlerror());
-#endif
+    const char *dlerr = dlerror();
+    if (dlerr) {
```
For the case where another thread sets `dlerror` at the wrong moment, I'd prefer keeping `!address` as the condition here. We should only look at `dlerr` if we actually got NULL.

grgalex replied:
Ok, I see your point.

Even though glibc's `dlerror()` is thread safe, it doesn't really matter to us.

So, let's flip the `if`s, and first check for NULL address, then for `dlerr`.

In this way, and given that our `#ifdefs` are not granular enough (esp. in this case where we are in an `#else`), we won't get bitten on a system where `dlerror()` is not thread-safe.

Of course, we could always print a wrong message, but that is just as possible even if we don't flip the `if`s.

WDYT?

encukou committed:
Write to a pipe instead of stderr

encukou committed:
Reorganize dlerror() code

- Only clear the error if we'll get it later
- De-duplicate the default message

encukou committed:
Use PyUnicode_DecodeLocale for the error

grgalex committed:
Clean up PyCFuncPtr_FromDll, uses #define USE_DLERROR

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
Clean up py_dl_sym, uses #define USE_DLERROR

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
test: Exercise 3 code paths, one for each C function we want to test

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
Add empty statement after label, placate C compiler

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
Move label to before #endif (WIN32)

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

encukou commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3774,19 +3794,52 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
+            PyErr_Clear();
+        }
+#endif
+#undef USE_DLERROR
```
nitpicks to make this easier to read:
- Consider moving this `#undef` after the final `#endif`, so the `#undef USE_DLERROR`...`#undef USE_DLERROR` region of code doesn't cross `{`...`}` and `#if`...`#endif` blocks
- Consider indenting `#ifdef USE_DLERROR` & its `#endif`.

encukou commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3774,19 +3794,52 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
+    if (address) {
+        goto dlsym_ok;
+    }
+    else {
```
Hm, am I missing something?
All the non-Windows error handling is in one block, so you shouldn't need the `goto` and label.
```suggestion
    if (!address) {
```

grgalex replied:
I think I was the one who missed something! Fixed.

encukou commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3774,19 +3794,52 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
+    else {
+#ifdef USE_DLERROR
+        // This assumes the error message is UTF-8 (or ASCII).
+        // Investigate if this can cause problems.
```
My suggestion left an outdated comment in, sorry!
```suggestion
```

grgalex committed:
Import ctypes, _ctypes inside test, so we don't get ImportError in windows systems

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

encukou reviewed (commented):
`_ctypes` does exist on Windows. But it will be missing if it fails to compile.

grgalex committed:
Revert "Import ctypes, _ctypes inside test, so we don't get ImportError in windows systems"

This reverts commit 8356cf48be2789f6a9589728915d5fb137b74900.

encukou commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,110 @@
+        retcode = subprocess.call(["gcc --version"],
+                                  stdout=subprocess.DEVNULL,
+                                  stderr=subprocess.DEVNULL,
+                                  shell=True)
```
`shell=True` isn't needed here; it's better to not use it.
```suggestion
        retcode = subprocess.call(["gcc", "--version"],
                                  stdout=subprocess.DEVNULL,
                                  stderr=subprocess.DEVNULL)
```

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -956,32 +956,51 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
+    // Investigate if this can cause problems.
+    const char *dlerr = dlerror();
+    if (dlerr) {
+        PyObject *message = PyUnicode_DecodeLocale(dlerr, "strict");
```
I think we want to use `surrogateescape` instead of raising a `UnicodeError`:
```suggestion
        PyObject *message = PyUnicode_DecodeLocale(dlerr, "surrogateescape");
```

grgalex replied:
Can you explain why?

ZeroIntensity replied:
`strict` will raise a `UnicodeError` if it doesn't like something when decoding, I don't think that's a good idea when it comes to decoding the locale. For precedent, we also typically use `surrogateescape` when using `DecodeLocale` in most of the core, as far as I can see.

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -956,32 +956,51 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
+            return NULL;
+        }
+        // Ignore errors from converting the message to str
+        PyErr_Clear();
```
This doesn't seem right, why? We might want to use `PyErr_WriteUnraisable` here.

grgalex replied:
I cannot argue for or against this, as I don't have the context.

Am I expected to get the context, or should I wait for someone else (the other reviewers) to answer?

ZeroIntensity replied:
Did @encukou tell you to do it this way? (If so, I'll yield.)

encukou replied:
I did, but before I saw the precedent of `surrogateescape` (in handling `errno`/`strerror` for `OSError`, in `errors.c`).
With `surrogateescape`, an exception raised here will should truly unexpected (e.g. a MemoryError); we can raise that instead of the generic ValueError.
```suggestion
        return NULL;
```

grgalex replied:
So, what's the verdict? I'm not 100% in sync.

ZeroIntensity replied:
Use `surrogateescape` and then propagate errors when decoding to the caller.

grgalex replied:
Done and pushed, PTAL to confirm in case I misunderstood something

ZeroIntensity replied:
Sorry, this PR got lost in notifications. I'll review this again tomorrow. If I forget, please ping me!

ZeroIntensity commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -956,32 +956,51 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
+
+    PyErr_Format(PyExc_ValueError,
+                    "symbol '%s' not found",
+                    name);
```
Align indentation
```suggestion
                "symbol '%s' not found",
                name);
```

ZeroIntensity reviewed (commented):
Sorry for dismissing the review--I wasn't aware of thread safety issues and whatnot before Petr brought them up.

That aside, in general, I'm not a fan of copying code between two files. Is it possible to unify these at all to prevent repeating? (e.g. an inline function in a header file?)

grgalex committed:
Remove unnecessary goto, fix preprocessor cmd indentation

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
test: Remove shell=True from gcc check, use separate arguments

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
Remove extra whitespace

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
test: Move imports to func, avoid ImportError on Windows

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
test: Explain why we moved imports to within test function

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

grgalex committed:
dlerror: Use surrogateescape error strategy for decoding locale

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

encukou commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -956,32 +956,47 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
+        if (message) {
+            PyErr_SetObject(PyExc_ValueError, message);
+            return NULL;
+        }
```
This will ignore exceptions from `PyUnicode_DecodeLocale` (since the `PyErr_Format` below will overwrite the current exception). If that's the intent, there should be an explicit `PyErr_Clear`.

```suggestion
            PyErr_SetObject(PyExc_ValueError, message);
            return NULL;
        }
        // Ignore errors from PyUnicode_DecodeLocale,
        // fall back to the generic error below.
        PyErr_Clear();
```
I still think that this (with `PyErr_Clear`) would be best: the generic `symbol '%s' not found` is more helpful than “sorry, we couldn't produce your error message”. (And unlike Python, a `KeyboardInterrupt` can't appear at an arbitrary point in C code -- we'd really be *only* ignoring `PyUnicode_DecodeLocale` errors.)

---

But if the intent is to propagate any exception from `PyUnicode_DecodeLocale` (instead of the `ValueError("symbol not found")`), that would be fine too. That would need a `return NULL` after the *if* body, though.

(same 2× below)

ZeroIntensity replied:
Yeah, but `symbol not found` might be very misleading to users, especially if the library actually has the symbol. Instead of clearing the error, could we just call `PyErr_WriteUnraisable` so the user at least sees *something*?

grgalex replied:
I agree with @ZeroIntensity that falling back to `symbol not found` may be misleading in some cases, especially if the cause of the error is not the absence of the symbol (or an IFUNC returning NULL), but something else is going on.

So, I think we should propagate the exception, and maybe accompany it with the non-decoded string, if that makes any sense. Would `PyErr_WriteUnraisable` help in this case?

encukou replied:
IMO, `symbol not found` is correct: if `dlsym` gives us NULL, we *can* say that Python did not find the symbol. (We'd like to say *why*, but if we fail to get that reason, a generic error is fine. It better be fine -- it's what Windows/Cygwin users will see.)
In other words: if an IFUNC tells us that the symbol is missing (NULL), well, that means we can't find it :)

But if you disagree: the error from `PyUnicode_DecodeLocale` is raisable. If it's helpful to the user, then raise it (i.e. return `NULL`).

grgalex replied:
Your point makes sense. From the scope of the callee (Python user), it's indistinguishable.

Let's not complicate things further.

Let's just use the fallback `symbol not found` error.

@ZeroIntensity what do you think?

ZeroIntensity replied:
I'm not the biggest fan of swallowing errors, but I'll defer to Petr. (That means, go ahead and commit his suggestion and I'll approve again.)

grgalex replied:
Done and pushed!

I also tried indenting the `#ifdef USE_DLERROR` / `#undef` according to https://github.com/python/cpython/pull/126555#discussion_r1838310417

encukou commented on Modules/_ctypes/callproc.c:
```diff
@@ -1623,13 +1623,35 @@ static PyObject *py_dl_sym(PyObject *self, PyObject *args)
     if (PySys_Audit("ctypes.dlsym/handle", "O", args) < 0) {
         return NULL;
     }
+    #ifdef __CYGWIN__
```
For consistency:
```suggestion
#undef USE_DLERROR
    #ifdef __CYGWIN__
```

encukou reviewed (commented):
This looks great, thank you!
Two more comments, and I hope that's it :)

(Assuming there'll be a later PR to remove the code duplication -- there's another issue for that.)

grgalex committed:
Update Modules/_ctypes/callproc.c

Co-authored-by: Petr Viktorin <encukou@gmail.com>

grgalex committed:
Clear DecodeLocale error, indent #ifdef, #undef

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

ZeroIntensity reviewed (approved):
I don't feel like nitpicking formatting will be useful, so LGTM (again)!

encukou committed:
Decref the message

encukou reviewed (approved):
If the buildbots like it, let's merge!
It's end of day for me, I'll get to it tomorrow. I pushed a refcounting fix directly to save time.

Thank you for the fix, and your patience, @grgalex

encukou commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,114 @@
+                f.write(FOO_C.replace('$DESCRIPTOR', str(pipe_w)))
+            args = ['gcc', '-fPIC', '-shared', '-o', dstname, srcname]
+            p = subprocess.run(args, capture_output=True)
+            self.assertEqual(p.returncode, 0, p)
```
IFUNC isn't just `gcc`-specific, it's also not available on some architectures. I guess this needs something like:
```suggestion
            if p.returncode != 0:
                # IFUNC is not available on all CPUs.
                if platform.machine() == 'x86_64':
                    # It should be suported here. We have a different problem.
                    p.check_returncode()
                else:
                    # IFUNC might not be supported on this machine.
                    self.skipTest(f"could not compile indirect function: {p}")
```

grgalex replied:
Done and pushed!

encukou reviewed (commented):
Well, not all buildbots liked it :(

encukou commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -0,0 +1,114 @@
+   Adapted from Michael Kerrisk's answer: https://stackoverflow.com/a/53590014.
+*/
+
+asm (".type foo, @gnu_indirect_function");
```
`asm` syntax is architecture-specific; the [cross-platform but GCC-only way to spell this](https://www.sourceware.org/binutils/docs/as/Type.html#Type) is:
```suggestion
asm (".type foo, STT_GNU_IFUNC");
```

grgalex replied:
From the documentation you linked:
```
The syntaxes supported are:

  .type <name> STT_<TYPE_IN_UPPER_CASE>
  [...]
  .type <name>,@<type>
  [...]
```

Therefore the two choices we have are:

1. `.type foo, @gnu_indirect_function`
2. `.type foo STT_GNU_IFUNC`

Notice the lack of comma in (2).

Your commit suggestion offers a hybrid variant, which strangely still works.

Nevertheless, these two seem equivalent to me. What makes the `STT_GNU_IFUNC` variant cross-platform?

encukou replied:
Oh, thanks for the catch!
```suggestion
asm (".type foo STT_GNU_IFUNC");
```

The docs also say:

> Because some of the characters used in these syntaxes (such as ‘@’ and ‘#’) are comment characters for some architectures, some of the syntaxes below do not work on all architectures. The first variant will be accepted by the GNU assembler on all architectures so that variant should be used for maximum portability, if you do not need to assemble your code with other assemblers.

Here's the failure from a Raspberry Pi: https://buildbot.python.org/api/v2/logs/11673079/raw_inline

grgalex replied:
Ah I see, so we both fixed something here :stuck_out_tongue:

grgalex committed:
Update Lib/test/test_ctypes/test_dlerror.py

Co-authored-by: Petr Viktorin <encukou@gmail.com>

grgalex committed:
test: Handle different architectures, where IFUNC not supported

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>

picnixz commented on Modules/_ctypes/callproc.c:
```diff
@@ -1623,13 +1623,40 @@ static PyObject *py_dl_sym(PyObject *self, PyObject *args)
-                               dlerror());
-        return NULL;
+    if (ptr)
+        return PyLong_FromVoidPtr(ptr);
```
```suggestion
    if (ptr) {
        return PyLong_FromVoidPtr(ptr);
    }
```

picnixz commented on Modules/_ctypes/callproc.c:
```diff
@@ -1623,13 +1623,40 @@ static PyObject *py_dl_sym(PyObject *self, PyObject *args)
+	#undef USE_DLERROR
+    PyErr_Format(PyExc_OSError,
+                 "symbol '%s' not found",
+                 name);
```
```C
    PyErr_Format(PyExc_OSError, "symbol '%s' not found", name);
```

I think this should fit on 80-chars.

grgalex replied:
Done!

picnixz commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -956,32 +956,50 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
+#undef USE_DLERROR
+    PyErr_Format(PyExc_ValueError,
+                 "symbol '%s' not found",
+                 name);
```
```C
    PyErr_Format(PyExc_ValueError, "symbol '%s' not found", name);
```

I think this should fit on 80-chars.

grgalex replied:
Done! Also fixed the other `PyErrFormat` on line 3822.

picnixz reviewed (approved):
One PEP-7 nitpick and two de-chop suggestions. I didn't write suggestions for the `PyErr_SetString` usages since I wasn't sure it fits on 80 chars.

grgalex committed:
Update Modules/_ctypes/callproc.c

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

grgalex committed:
Don't break when under 80 chars

Signed-off-by: Georgios Alexopoulos <grgalex@ba.uoa.gr>
